pkg/controller/service/service_controller.go

```diff
@@ -138,11 +131,11 @@ func New(
 			UpdateFunc: func(old, cur interface{}) {
 				oldSvc, ok1 := old.(*v1.Service)
 				curSvc, ok2 := cur.(*v1.Service)
-				if ok1 && ok2 && s.needsUpdate(oldSvc, curSvc) {
+				if ok1 && ok2 && (s.needsUpdate(oldSvc, curSvc) || isDeletionCandidate(curSvc)) {
```

Comment:
What if users deletes the service, and then updates the service from "loadbalancer" to "clusterIP"? Then it's possible that the service controller only receives one update event showing the service is being deleted and it's of the type "clusterIP". Will the controller leave an orphaned load balancer?

Reply:
@caesarxuchao are you referring to this:

1. User deletes a LB service
2. As it's deleting they recreate it w/ a different type (cluster ip for example)

What happens?

The update request should fail. 

The deletion timestamp would be set on the service which means the apiserver will only allow updates to the finalizers, so you wouldn't be able to re-create/update it until the deletion is complete.

Reply:
> As it's deleting they recreate it w/ a different type (cluster ip for example)

I was referring to "update" specifically. "Recreate" should be rare, a user would have to intentionally remove the finalizer to complete the deletion, and then recreate the object. Otherwise it will be a 409 error.

> The deletion timestamp would be set on the service which means the apiserver will only allow updates to the finalizers, so you wouldn't be able to re-create/update it until the deletion is complete.

Is this true? I don't think apiserver enforces that. Maybe service has specific validation logic to enforce it. Could you double check?

Reply:
> What if users deletes the service, and then updates the service from "loadbalancer" to "clusterIP"? Then it's possible that the service controller only receives one update event showing the service is being deleted and it's of the type "clusterIP".

Thanks for chiming in! This is an interesting use case. As of current implementation, even if user is able to update the service to non-loadbalancer type after deletion, service controller would still attempt to remove corresponding loadbalancer because of the attached finalizer. Upon a service deletion update, controller only cares about whether it has the LB finalizer, but not which type it has.

Reply:
Thanks @MrHohn. That clears my concerns.

Reply:
@caesarxuchao 

> Is this true? I don't think apiserver enforces that. Maybe service has specific validation logic to enforce it. Could you double check?

Either i'm going crazy or i remember reading that some where in the docs. Now to figure out where I read it... 

As @MrHohn said it won't matter because the load balancer finalizer would remain on the service until deletion. Worst case is you have temporary orphan resources if a delete failed and you updated the svc enough that the previous and current services both don't trigger a LB change logic. Even then though that would be cleaned up when the service is deleted since the finalizer would exist. 

We could add a check for that case to speed up cleaning up the resources. Basically `if finalizer exists & not a load balancer service then run delete logic` I'd prefer to do that in a separate PR so we can get this landed in time tho.